### PR TITLE
DEV: Add rake command to help detect dead settings

### PR DESCRIPTION
**Pull request by oblakeerickson**
Some Site Settings may still exist but are no longer being used in the
core discourse code or in related plugins. This rake task will help
identify any unused (aka: dead) settings by using the `rg` command to
search for them.

You can execute the rake task by using this command:

`LOAD_PLUGINS=1 bin/rails "site_settings:find_dead"`

### CODE REVIEW

File: app/services/site_settings_task.rb [hunk 1]
```diff
@@ -38,4 +38,35 @@ def self.import(yml)
...
+  def self.rg_installed?
+    !`which rg`.strip.empty?
+  end
+
+  def self.directory_path(directory_name)
+    File.expand_path(File.join(Dir.pwd, "..", directory_name))
+  end
+
+  def self.directories_to_check
+    %w[all-the-themes all-the-custom-themes all-the-plugins all-the-custom-plugins]
```
Comment (martin-brennan) [inline app/services/site_settings_task.rb:56]: Only thing I would add is being able to specify a directory (using an ENV) where all these repos are kept...I keep these totally separate from the core discourse repo in `$HOME/allthe`

File: app/services/site_settings_task.rb [hunk 2]
```diff
@@ -38,4 +38,35 @@ def self.import(yml)
     end
     [log, counts]
   end
+
+  def self.names
+    site_settings = SiteSetting.all_settings(include_hidden: true)
```
Comment (martin-brennan) [inline app/services/site_settings_task.rb:43]:
```suggestion
    SiteSetting.all_settings(include_hidden: true).map { |site_setting| site_setting[:setting].to_s }
```

File: lib/tasks/site_settings.rake [hunk 3]
```diff
@@ -34,3 +34,53 @@ task "site_settings:import" => :environment do
...
+    puts "Please install ripgrep to use this command"
+    exit 1
+  end
+
+  puts "Checking #{setting_names.count} settings in these directories:"
+  puts directories
+  puts
+
+  setting_names.each do |setting_name|
+    count = 0
```
Comment (martin-brennan) [inline lib/tasks/site_settings.rake:63]: No need for this, since you override it straight away anyway

File: lib/tasks/site_settings.rake [hunk 4]
```diff
@@ -34,3 +34,53 @@ task "site_settings:import" => :environment do
...
+    count = SiteSettingsTask.rg_search_count("SiteSetting.#{setting_name}", directories[0])
```
Comment (martin-brennan) [inline lib/tasks/site_settings.rake:64]: Total nitpick, but you could use rails magic `directories.first` :P

File: lib/tasks/site_settings.rake [hunk 5]
```diff
@@ -34,3 +34,53 @@ task "site_settings:import" => :environment do
...
+    count =
+      SiteSettingsTask.rg_search_count("siteSettings.#{setting_name}", directories[0]) if count == 0
+    directories.each do |directory|
+      count = SiteSettingsTask.rg_search_count(setting_name, directory) if count == 0
+    end
+    if count == 0
```
Comment (martin-brennan) [inline lib/tasks/site_settings.rake:70]: Same, could use `count.zero?` here

**Review by martin-brennan — APPROVED** [review]
Hope we don't get banned from TikTok for all these uses of `dead`, consider changing to `unalive` (joking, joking).

Looks great, mostly just minor changes!

File: app/services/site_settings_task.rb (hunk 1 above)
Comment (oblakeerickson) [reply app/services/site_settings_task.rb:56]: Ya everyone's setups are different, good call, I'll add that.

File: lib/tasks/site_settings.rake (hunk 4 above)
Comment (oblakeerickson) [reply lib/tasks/site_settings.rake:64]: lol, that it exactly what I was doing in the rails console playing around with this

File: lib/tasks/site_settings.rake (hunk 5 above)
Comment (oblakeerickson) [reply lib/tasks/site_settings.rake:70]: Ya, I like it.

**Pull request merged**